File: 2017/Python/01.py

def captcha_sum(seq):
  return sum([int(num) for i, num in enumerate(seq) 
      if seq[i+1 if i < len(seq)-1 else 0] == num])

def captcha_sum_halfway(seq):
  return sum([int(num) for i, num in enumerate(seq) 
      if seq[int((i + len(seq)/2) % len(seq))] == num])



assert(captcha_sum("1122") == 3)
assert(captcha_sum("1111") == 4)
assert(captcha_sum("1234") == 0)
assert(captcha_sum("91212129") == 9)

assert(captcha_sum_halfway("1212") == 6)
assert(captcha_sum_halfway("1221") == 0)
assert(captcha_sum_halfway("123425") == 4)
assert(captcha_sum_halfway("123123") == 12)
assert(captcha_sum_halfway("12131415") == 4)

with open("../inputs/01.txt") as f:
  puzzle_seq = f.read().strip()
  print(captcha_sum(puzzle_seq))
  print(captcha_sum_halfway(puzzle_seq))

File: 2017/C++/01.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <cstddef>

// TODO: Extract this to some utils library?
std::vector<std::string> read_lines(const std::string &filename)
{
    std::vector<std::string> vec;
    std::ifstream f(filename);
    std::string s;
    while (std::getline(f, s)) {
        vec.push_back(s);
    }
    return vec;
}

int char_to_int(char c)
{
    int n = c - '0';
    return n;
}

int captcha_sum_naive(std::string str, bool halfway=false)
{
    int sum = 0;
    for (std::size_t i = 0; i < str.size(); ++i) {
        std::size_t i_next;
        if (halfway)
            i_next = (i + str.size() / 2) % str.size();
        else
            i_next = i < (str.size() - 1) ? i + 1 : 0;
        int num = char_to_int(str[i]);
        int next_num = char_to_int(str[i_next]);
        if (num == next_num)
            sum += num;
    }

    return sum;
}

int main()
{
    auto v = read_lines("../../inputs/01.txt");
    for (const auto& e : v) {
        std::cout << captcha_sum_naive(e) << std::endl;
        std::cout << captcha_sum_naive(e, true) << std::endl;
    }

    return 0;
}
